### PR TITLE
Add option to host an "echo" server

### LINKED ISSUE #73: Echo benchmark
Maybe interesting to do a benchmark that is close to the best echo server we can implement.
Such a server would fully run on the iothread and batch all reads and writes.
It doesn't use ASP.NET and ThreadPool, just plain Socket.

For benchmarking purposes, instead of implementing echo, we can implement a hard-coded JSON benchmark. I think the length of request for json is fixed. So each time we receive that amount of bytes, we can send a fixed answer.

cc @antonfirsov @adamsitnik

### PR DESCRIPTION
**Pull request by antonfirsov**
Resolves #73.

Dumb as hell, can't handle concurrent connections.

### CODE REVIEW

File: test/web/RawSocketHost.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            Console.WriteLine($"Raw server listening on {_serverEndpoint}");
+            
+            byte[] receiveBuffer = new byte[BufferSize];
+            byte[] sendBuffer = Encoding.ASCII.GetBytes(Response);
+
+            while (true)
+            {
+                using Socket handler = await listener.AcceptAsync();
+
+                while (true)
```
Comment (tmds) [inline test/web/RawSocketHost.cs]: Can you add a class that handles a single client?
```cs
Socket clientSocket = await listener.AcceptAsync();
new ClientConnection(clientSocket, <options>).HandleClient();
```

`ClientConnection` should use 2 `SocketAsyncEventArgs`, one for receiving, and one for sending.
We can support some of the options, like `c=inline`, by setting properties on the `SocketAsyncEventArgs`.

**Comment by antonfirsov** [comment]
@tmds does it look better now? any further suggestions?

File: test/web/RawSocketHost.cs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+                _options.SocketContinuationScheduler == SocketContinuationScheduler.ThreadPool;
+
+            using ClientConnectionHandler clientHandler = 
+                new ClientConnectionHandler(deferSends, deferReceives, runContinuationsAsynchronously);
+            
+            while (true)
+            {
+                using Socket handlerSocket = await listener.AcceptAsync();
+
+                clientHandler.HandleClient(handlerSocket);
```
Comment (tmds) [inline test/web/RawSocketHost.cs]: We want to new up a ClientConnectionHandler for each Socket, that way we can handle multiple sockets.
The ClientConnectionHandler will own the Socket, it doesn't need to be IDisposable.

File: test/web/RawSocketHost.cs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        {
+            private Socket _socket;
+            private readonly byte[] _sendBuffer = Encoding.ASCII.GetBytes(Response);
+            private readonly byte[] _receiveBuffer = new byte[BufferSize];
+
+            private readonly SocketAsyncEventArgs _receiveArgs;
+            private readonly SocketAsyncEventArgs _sendArgs;
+            private readonly ManualResetEventSlim _sendEvent = new ManualResetEventSlim(true);
+            private readonly ManualResetEventSlim _receiveEvent = new ManualResetEventSlim();
+
```
Comment (tmds) [inline test/web/RawSocketHost.cs]: We don't need the `ManualResetEventSlim`.
We should be able to do something like:

```cs
ClientConnectionHandler()
{
      _sendArgs.Completed += (s, a) => ReceiveRequest();
      _receiveArgs.Completed += (s, a) => OnReceive();
}

HandleClient() => ReceiveRequest()

ReceiveRequest()
{
	try
	{
                _receiveArgs.SetBuffer(_receiveBuffer, _bytesReceived);
		if (!_socket.ReceiveAsync(_receiveArgs))
		{
		    OnReceive();
		}
	}
	catch
	{
		Cleanup();
	}
}

OnReceive()
{
	int bytesTransferred = _receiveArgs.BytesTransferred;
	SocketError error = _receiveArgs.SocketError;

	if (error == SocketError.Success && count > 0)
	{
		_bytesReceived += bytesTransferred;
		if (RequestComplete(...))
		{
			_bytesReceived = 0;
			SendResponse();
		}
                else
                {
                        ReceiveRequest();
                }
	}
	else
	{
		Cleanup();
	}
}

SendResponse()
{
	try
	{
		if (!_socket.SendAsync(_sendArgs))
		{
		    ReceiveRequest();
		}
	}
	catch
	{
		Cleanup();
	}	
}

Cleanup()
{
	_socket.Dispose();
}
```

File: test/web/RawSocketHost.cs (hunk 2 above)
Comment (tmds) [reply test/web/RawSocketHost.cs]: We may need to keep the handler objects in a collection. Though I expect the on-going async operations will keep them rooted also.

**Comment by antonfirsov** [comment]
@tmds I kept my handling logic sequential ("echo-serverish"). Based on your comments, it looks like you want to enable handling connections in parallel, right? Makes sense if we want to test various scheduling modes.

**Comment by tmds** [comment]
> Based on your comments, it looks like you want to enable handling connections in parallel, right?

Yes. And we're in full control of how many epoll threads there are and how many clients connect. So we can still connect a single client and run that on a single epoll thread.

> I kept my handling logic sequential ("echo-serverish").

I guess this is about my comment to refactor `ClientConnectionHandler`? The suggested refactoring is sequential:

When a send is done, start a receive:
```cs
_sendArgs.Completed += (s, a) => ReceiveRequest();
```
When a request is complete, send a response:
```cs
		if (RequestComplete(...))
		{
			_bytesReceived = 0;
			SendResponse();
		}
```

`ManualResetEventSlim` will give some overhead, we don't need it.

**Comment by antonfirsov** [comment]
@tmds just pushed your suggestions, works smoothly on my machine.

**Comment by tmds** [comment]
LGTM! Thanks!

**Pull request merged**